torch/nn/parallel/data_parallel.py

```diff
@@ -123,6 +124,11 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
         if output_device is None:
             output_device = device_ids[0]
 
+        if not all(t.is_cuda and t.device.index == device_ids[0]
+                   for t in chain(module.parameters(), module.buffers())):
+            raise RuntimeError("module must have its parameters and buffers "
```

Comment:
What would go wrong if it is on device[1]?

Reply:
It hits the following error in `torch._C._broadcast_coalesced`. It matches the behavior described in [docs](thttps://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/data_parallel.py#L55-L57), and it might relates to the [guarantee](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/data_parallel.py#L64-L70) it provided for in-place update.

```
In [5]: dp = DataParallel(nn.Linear(2, 3).cuda(1))
                                                                               
In [6]: dp(torch.rand(2, 2))
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-d3b833c877f9> in <module>     
----> 1 dp(torch.rand(2, 2))
----> 1 dp(torch.rand(2, 2))

~/project/pytorch/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    489             result = self._slow_forward(*input, **kwargs)
    490         else:
--> 491             result = self.forward(*input, **kwargs)
    492         for hook in self._forward_hooks.values():
    493             hook_result = hook(self, input, result)

~/project/pytorch/torch/nn/parallel/data_parallel.py in forward(self, *inputs, **kwargs)
    146         if len(self.device_ids) == 1:
    147             return self.module(*inputs[0], **kwargs[0])
--> 148         replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
    149         outputs = self.parallel_apply(replicas, inputs, kwargs)
    150         return self.gather(outputs, self.output_device)

~/project/pytorch/torch/nn/parallel/data_parallel.py in replicate(self, module, device_ids)
    151 
    152     def replicate(self, module, device_ids):
--> 153         return replicate(module, device_ids)
    154 
    155     def scatter(self, inputs, kwargs, device_ids):

~/project/pytorch/torch/nn/parallel/replicate.py in replicate(network, devices, detach)
     11     params = list(network.parameters())
     12     param_indices = {param: idx for idx, param in enumerate(params)}
---> 13     param_copies = Broadcast.apply(devices, *params)
     14     if len(params) > 0:
     15         param_copies = [param_copies[i:i + len(params)]

~/project/pytorch/torch/nn/parallel/_functions.py in forward(ctx, target_gpus, *inputs)
     19         ctx.num_inputs = len(inputs)
     20         ctx.input_device = inputs[0].get_device()
---> 21         outputs = comm.broadcast_coalesced(inputs, ctx.target_gpus)
     22         non_differentiables = []
     23         for idx, input_requires_grad in enumerate(ctx.needs_input_grad[1:]):

~/project/pytorch/torch/cuda/comm.py in broadcast_coalesced(tensors, devices, buffer_size)
     38         corresponding to indices from ``devices``.
     39     """
---> 40     return torch._C._broadcast_coalesced(tensors, devices, buffer_size)
     41 
     42 

RuntimeError: all tensors must be on devices[0]
```

Reply:
Okay thanks! The in-place guarantee shouldn't matter, as it can share memory with the replica on device[1], but let's keep this constraint as it is already here.